python/lsst/ts/interlockmonitor/thermal/controller.py

```diff
@@ -0,0 +1,249 @@
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = [
+    "MockThermalScanner",
+    "ThermalScanner",
+    "ThermalScannerServer",
+    "ThermalScannerMockClient",
+]
+
+import asyncio
+import logging
+import os
+import pty
+
+import aioserial
+from lsst.ts import tcpip, utils
+
+
+class MockThermalScanner:
+    """Implement a mock thermal scanner.
+
+    Attributes
+    ----------
+    temperature : `list` of `float`
+        The array of current temperature data.
+    serial : `None`
+        The serial commander.
+    log : `logging.Logger`
+        The log instance.
+    """
+
+    def __init__(self) -> None:
+        self.temperatures = [0] * 6
+        self.serial = None
+        self.log = logging.getLogger(__name__)
+        self.log.info("MockThermalScanner created.")
+
+    def start_port(self):
+        """Open the serial port and return name of client pty."""
+        self.master, self.slave = pty.openpty()
+        self.log.info("Create ports.")
+        self.log.info(f"{os.ttyname(self.master)}")
+        self.serial = aioserial.AioSerial(
+            port=os.ttyname(self.master), baudrate=19200, cancel_write_timeout=10
+        )
+        self.serial.fd = self.master
+        self.log.info("Open serial.")
+        return os.ttyname(self.slave)
+
+    async def reply(self):
+        """Generate reply and send data over port."""
+        while True:
+            for i in range(0, 5):
+                if i == 4:
+                    reply = f"C0{i+1}={self.temperatures[i]:03.4f}".encode("ISO-8859-1")
+                    bytes_written = await self.serial.write_async(reply)
+                    self.log.info(f"{bytes_written=}")
+                    bytes_written = await self.serial.write_async(
+                        aioserial.CR + aioserial.LF
+                    )
+                    self.log.info(f"{bytes_written=}")
+                else:
+                    reply = f"C0{i+1}={self.temperatures[i]:03.4f},".encode(
+                        "ISO-8859-1"
+                    )
+                    self.log.info(f"{reply=}")
+                    bytes_written = await self.serial.write_async(reply)
+                    self.log.info(f"{bytes_written=}")
+                await asyncio.sleep(0.175)
```

Comment:
Why do you wait 0.175 seconds to write each value but wait 5 times before writing the terminator? Is that what the real system does, and does it matter? I would be a lot more common to write the whole string all at once (all 5 values), terminator and all, and wait 5x as long).

Reply:
According to the document describing the messages, yes it waits approximately 0.175 seconids per each channel, not sure if it matters or not but it mentions it explicitly in the documentation. I have yet to get to the real device due to the RPi being offline for some tests that are being run in the calibration lab. I'll change it if the device behaves differently or it may not matter.